File: Ruby/conway.rb
class Conway
    attr_accessor :height, :width, :grid

    def initialize(height, width)
        self.height = height
        self.width = width
        self.grid = Array.new(height) { Array.new(width, 0) }
    end

    def query(x, y)
        return self.grid[x % self.height][y % self.width]
    end

    def assign(x, y, state)
        self.grid[x % self.height][y % self.width] = state
    end

    def count_neighbors(x, y)
        n  = self.query(x + 1, y    )
        ne = self.query(x + 1, y + 1)
        e  = self.query(x,     y + 1)
        se = self.query(x - 1, y + 1)
        s  = self.query(x - 1, y    )
        sw = self.query(x - 1, y - 1)
        w  = self.query(x,     y - 1)
        nw = self.query(x + 1, y - 1)
        neighbor_states = [n, ne, e, se, s, sw, w, nw]
        return neighbor_states.sum
    end

    def print_grid
        for i in 0..height - 1
            for j in 0..width - 1
                putc self.grid[i][j] == 1 ? "#" : "."
            end
            puts
        end
    end

    def game_logic(state, neighbors)
        case state
        when 1
            if neighbors < 2
                return 0
            end
            if neighbors > 3
                return 0
            end
            return state
        when 0
            if neighbors == 3
                return 1
            end
            return state
        end
    end

    def step_cell(x, y)
        state = self.query(x, y)
        neighbors = self.count_neighbors(x, y)
        next_state = self.game_logic(state, neighbors)
        return next_state
    end

    def simulate
        new_grid = Array.new(height) { Array.new(width, 1) }

        for i in 0..height - 1
            for j in 0..width - 1
                new_grid[i][j] = self.step_cell(i, j)
            end
        end

        self.grid = new_grid
    end

    def clear_screen
        puts `clear`
        # puts "\e[H\e[2J"
    end

end

g = Conway.new(10, 20)

g.assign(0, 3, 1)
g.assign(1, 4, 1)
g.assign(2, 2, 1)
g.assign(2, 3, 1)
g.assign(2, 4, 1)

g.print_grid

while true
    g.clear_screen
    g.simulate
    g.print_grid
    sleep(0.1)
end

# ruby conway.rb

File: Perl/conway.pl
use POSIX;
use Time::HiRes;

sub new_grid {
    my ($height, $width) = @_;
    my @grid = ();

    for ($i = 0; $i < $height; $i = $i + 1) {
        for ($j = 0; $j < $width; $j = $j + 1) {
            $grid[$i][$j] = 0;
        }
    }

    return @grid;
}

sub query {
    my ($x, $y) = @_;
    return $g[fmod($x, $h)][fmod($y, $w)];
}

sub assign {
    my ($x, $y, $state) = @_;
    $g[fmod($x, $h)][fmod($y, $w)] = $state;
}

sub count_neighbors {
    my ($x, $y) = @_;
    my $n  = query($x + 1, $y    );
    my $ne = query($x + 1, $y + 1);
    my $e  = query($x,     $y + 1);
    my $se = query($x - 1, $y + 1);
    my $s  = query($x - 1, $y    );
    my $sw = query($x - 1, $y - 1);
    my $w  = query($x,     $y - 1);
    my $nw = query($x + 1, $y - 1);
    return $n + $ne + $e + $se + $s + $sw + $w + $nw;
}

sub print_grid {
    foreach my $row (@g) {
        foreach my $col (@$row) {
            print $col == 0 ? "." : "#";
        }
        print "\n";
    }
}

sub game_logic {
    ($state, $neighbors) = @_;
    if ($state == 1) {
        if ($neighbors < 2) {
            return 0;
        }
        if ($neighbors > 3) {
            return 0;
        }
        return $state;
    } else {
        if ($neighbors == 3) {
            return 1;
        }
        return $state;
    }
}

sub step_cell {
    my ($x, $y) = @_;
    my $state = query($x, $y);
    my $neighbors = count_neighbors($x, $y);
    my $next_state = game_logic($state, $neighbors);
    return $next_state;
}

sub simulate {
    my @new_g = new_grid($h, $w);

    for ($i = 0; $i < $h; $i = $i + 1) {
        for ($j = 0; $j < $w; $j = $j + 1) {
            $new_g[$i][$j] = step_cell($i, $j);
        }
    }

    @g = @new_g;
}

sub clear_screen {
    print "\x1B[2J";
}

our $h = 10;
our $w = 20;
our @g = new_grid($h, $w);

assign(0, 3, 1);
assign(1, 4, 1);
assign(2, 2, 1);
assign(2, 3, 1);
assign(2, 4, 1);

print_grid;

while (true) {
    clear_screen;
    simulate;
    print_grid;
    Time::HiRes::sleep(0.1);
}

# perl conway.pl
